### ISSUE TITLE
FP16 model silently returns wrong outputs with default specialization on iPhone 17 Pro; CPU-only, GPU-preferred and FP32 are correct

### ISSUE THREAD

**Issue by glenn-jocher** [apple/coreai-torch#115]
### Summary

An FP16 `.aimodel` returns wrong outputs on an iPhone 17 Pro when loaded with `SpecializationOptions.default` (or with `preferredComputeUnitKind: .neuralEngine`). The same asset with the same input is correct with `.cpuOnly` and with `preferredComputeUnitKind: .gpu`, and the FP32 export of the same model is correct with every option. There is no error, warning or NaN: the model loads, runs fast, and silently returns different numbers, with every confidence equal to `0.0`.

The model is the YOLO26n pose model exported with its end-to-end head (`topk` + `gather` in the graph, output `(1, 300, 57)`). The YOLO26n detect model exported the same way (`(1, 300, 6)`) is correct on the same device, so this is specific to the pose graph rather than to `topk` in general.

### Environment

- iPhone 17 Pro (`iPhone18,1`), iOS 27.0
- Xcode 27.0 (27A266a), iOS 27.0 SDK
- Export host: macOS 26.7, Apple silicon, Python 3.12, `coreai-core==1.0.0b2`, `coreai-torch==0.4.2`, `torch==2.14.0`, `ultralytics==8.4.155`

### Export

```bash
pip install ultralytics==8.4.155 coreai-torch==0.4.2
yolo export model=yolo26n-pose.pt format=coreai quantize=16 nms=False imgsz=640  # FP16, wrong on device
yolo export model=yolo26n-pose.pt format=coreai quantize=32 nms=False imgsz=640  # FP32, correct
yolo export model=yolo26n.pt      format=coreai quantize=16 nms=False imgsz=640  # FP16 detect control, correct
```

`yolo26n-pose.pt` and `yolo26n.pt` download automatically. The export path is `torch.export.export` → `run_decompositions(coreai_torch.get_decomp_table())` → `TorchConverter().add_exported_program(...)` → `to_coreai()` → `optimize()` → `save_asset()`, with the model and example input converted to half precision first for `quantize=16` ([source](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/export/coreai.py)).

### Reproduction

Bundle the three `.aimodel` directories in an iOS app and call `compare(url)` for each. It feeds a fixed pseudo-random input, so both runs see identical pixels, and compares `.cpuOnly` with `.default`.

```swift
// Minimal Core AI reproducer: same .aimodel, same input, CPU-only vs default specialization.
import CoreAI
import Foundation

@available(iOS 27.0, macOS 27.0, *)
func run(_ url: URL, _ options: SpecializationOptions) async throws -> [Float] {
  let model = try await AIModel(contentsOf: url, options: options)
  let function = try model.loadFunction(named: "main")!
  let name = function.descriptor.inputNames[0]
  guard case .ndArray(let input)? = function.descriptor.inputDescriptor(of: name) else { fatalError() }
  var seed: UInt32 = 1  // fixed LCG so both runs see identical pixels in 0-1
  let count = input.shape.reduce(1, *)
  let pixels = (0..<count).map { _ -> Float in
    seed = seed &* 1_664_525 &+ 1_013_904_223
    return Float(seed >> 8) / Float(1 << 24)
  }
  let array = input.scalarType == .float16
    ? NDArray(scalars: pixels.map { Float16($0) }, shape: input.shape)
    : NDArray(scalars: pixels, shape: input.shape)
  var outputs = try await function.run(inputs: [name: array])
  let out = outputs.remove(function.descriptor.outputNames[0])!.ndArray!
  if out.scalarType == .float16 {
    let span = out.view(as: Float16.self).contiguousElements!
    return (0..<span.count).map { Float(span[$0]) }
  }
  let span = out.view(as: Float.self).contiguousElements!
  return (0..<span.count).map { span[$0] }
}

@available(iOS 27.0, macOS 27.0, *)
func compare(_ url: URL) async {
  do {
    let cpu = try await run(url, .cpuOnly)
    let def = try await run(url, .default)
    let width = url.lastPathComponent.contains("pose") ? 57 : 6  // (1, 300, width): x1 y1 x2 y2 conf cls [kpts]
    let conf = { (o: [Float]) in stride(from: 4, to: o.count, by: width).map { o[$0] } }
    let diff = zip(cpu, def).map { abs($0 - $1) }.max()!
    print("REPRO \(url.lastPathComponent): max|cpu-default|=\(diff) nan(default)=\(def.filter(\.isNaN).count)")
    print("REPRO   cpu     top conf \(conf(cpu).sorted(by: >).prefix(5).map { String(format: "%.4f", $0) })")
    print("REPRO   default top conf \(conf(def).sorted(by: >).prefix(5).map { String(format: "%.4f", $0) })")
    print("REPRO   cpu     row0 \(cpu.prefix(6).map { String(format: "%.3f", $0) })")
    print("REPRO   default row0 \(def.prefix(6).map { String(format: "%.3f", $0) })")
  } catch { print("REPRO \(url.lastPathComponent) failed: \(error)") }
}
```

### Observed

```
REPRO yolo26n-pose.aimodel: max|cpu-default|=665.0 nan(default)=0
REPRO   cpu     top conf ["0.0023", "0.0015", "0.0015", "0.0011", "0.0009"]
REPRO   default top conf ["0.0000", "0.0000", "0.0000", "0.0000", "0.0000"]
REPRO   cpu     row0 ["577.000", "11.625", "640.500", "509.000", "0.002", "0.000"]
REPRO   default row0 ["0.158", "-0.406", "9.422", "22.016", "0.000", "0.000"]
REPRO yolo26n-pose-fp32.aimodel: max|cpu-default|=0.0029296875 nan(default)=0
REPRO   cpu     top conf ["0.0024", "0.0015", "0.0015", "0.0013", "0.0009"]
REPRO   default top conf ["0.0024", "0.0015", "0.0015", "0.0013", "0.0009"]
REPRO   cpu     row0 ["576.844", "12.519", "640.390", "508.734", "0.002", "0.000"]
REPRO   default row0 ["576.844", "12.519", "640.390", "508.734", "0.002", "0.000"]
REPRO yolo26n.aimodel: max|cpu-default|=640.875 nan(default)=0
REPRO   cpu     top conf ["0.3381", "0.0546", "0.0534", "0.0222", "0.0189"]
REPRO   default top conf ["0.3354", "0.0540", "0.0529", "0.0218", "0.0191"]
REPRO   cpu     row0 ["0.750", "-1.000", "638.500", "641.000", "0.338", "16.000"]
REPRO   default row0 ["0.000", "-0.375", "638.500", "640.000", "0.335", "16.000"]
```

For the FP16 pose asset, `.default` returns boxes that are not in input-pixel space and a confidence of exactly `0.0` in all 300 rows. The detect control's large `max|cpu-default|` comes from near-tied rows changing order between lanes; its confidences agree to FP16 precision.

With a real photograph (`bus.jpg`, letterboxed to 640x640) the same asset gives 4 people at 0.886 / 0.875 / 0.851 / 0.305 with `.cpuOnly` and no detections with `.default`.

### What was ruled out

| Variant | Result on device |
| --- | --- |
| FP16 pose, `.cpuOnly` | correct |
| FP16 pose, `preferredComputeUnitKind: .gpu` | correct (4 detections on `bus.jpg`) |
| FP16 pose, `.default` | **wrong**, all confidences `0.0` |
| FP16 pose, `preferredComputeUnitKind: .neuralEngine` | **wrong**, all confidences `0.0` |
| FP32 pose, `.default` | correct |
| FP16 pose exported without the end-to-end head (`nms=None`, output `(1, 56, 8400)`), `.default` | correct |
| FP16 detect with the end-to-end head, `.default` | correct |

So the failure needs FP16, the lane `.default` picks on this device, and the pose end-to-end postprocess, which, unlike detect, also gathers the keypoint channels for the selected rows.

### Expected

Either the same result as `.cpuOnly` within FP16 tolerance, or a load-time error or fallback if the selected lane cannot run the graph correctly. A silent wrong answer is the worst case for an application, because nothing can detect it at runtime.

**Comment by cdeil** [comment]
I hit the same failure with a DETR-style detector (RF-DETR) and narrowed it to `topk`: on the Neural Engine lane a float16 `topk` returns **16-bit indices packed into its int32 output**, two per element. Everything that consumes the indices then reads the wrong rows.

The smallest case I have: `topk(arange(8), k=4)` has exactly one answer, `[7, 6, 5, 4]` (coreai-torch 0.4.2, coreai-core 1.0.0b2, torch 2.14.0, macOS 27.0 26A428, M5 Pro):

```python
import asyncio, shutil
from pathlib import Path

import torch
import coreai.runtime as rt
from coreai_torch import TorchConverter, get_decomp_table


class TopK(torch.nn.Module):
    def forward(self, scores):
        return torch.topk(scores, 4, dim=1)[1].to(torch.int32)


scores = torch.arange(8, dtype=torch.float16)[None]
ep = torch.export.export(TopK(), (scores,)).run_decompositions(get_decomp_table())
converter = TorchConverter().add_exported_program(
    ep, input_names=["scores"], output_names=["indices"]
)
program = converter.to_coreai()
program.optimize()
path = Path("topk.aimodel")
shutil.rmtree(path, ignore_errors=True)
program.save_asset(path)


async def run(options):
    model = await rt.AIModel.load(path, options)
    outputs = await model.load_function("main")({"scores": rt.NDArray(scores.numpy())})
    return outputs["indices"].numpy().ravel()


prefer = rt.SpecializationOptions.from_preferred_compute_unit_kind
lanes = {
    "cpu": rt.SpecializationOptions.cpu_only(),
    "gpu": prefer(rt.ComputeUnitKind.gpu()),
    "neural_engine": prefer(rt.ComputeUnitKind.neural_engine()),
}
for name, options in lanes.items():
    indices = asyncio.run(run(options))
    print(name, indices.tolist(), [hex(i) for i in indices])
```

```text
cpu [7, 6, 5, 4] ['0x7', '0x6', '0x5', '0x4']
gpu [7, 6, 5, 4] ['0x7', '0x6', '0x5', '0x4']
neural_engine [393223, 262149, 131075, 1] ['0x60007', '0x40005', '0x20003', '0x1']
```

The Neural Engine result is the complete descending order 7, 6, …, 0 as 16-bit values read two at a time. So the lane seems to sort the whole axis with 16-bit indices, and the first `k` int32 words of that buffer are returned. With `torch.topk(scores.float(), ...)` inside the same float16 graph, the Neural Engine returns `[7, 6, 5, 4]`, and that is the workaround I use. With 576 scores and `k=300`, as in the detector, the largest "index" is 37,683,739.

**RF-DETR.** It selects queries with a `topk` over 576 encoder tokens followed by a `gather`. Float16 on the Neural Engine lane returned 0 of 39 detections on five test images before the workaround and 39 of 39 after it. On an M4 iPad and an A15 iPhone 13 (iPadOS/iOS 27.0, 24A437), `SpecializationOptions.default` runs float16 graphs on this lane. In an all-process Instruments "Core AI" trace, the Neural Engine is busy during those inferences and the GPU during float32 ones. All other Neural Engine activity in that trace falls within the app's own model loading. The default path therefore returned no detections; with the float32 `topk` it matches the CPU.

**Your pose model.** I ran your export (Ultralytics 8.4.158, `format="coreai"`, `quantize=16`, `nms=False`, torch 2.13.0) on the Mac's Neural Engine lane with `bus.jpg`:

```text
pose               cpu            detections>0.5: 3  top conf [0.884, 0.875, 0.852, 0.286]
pose               neural_engine  detections>0.5: 0  top conf [0.0, 0.0, 0.0, 0.0]
pose, fp32 topk    neural_engine  detections>0.5: 3  top conf [0.887, 0.875, 0.852, 0.298]
detect             neural_engine  detections>0.5: 5  top conf [0.927, 0.915, 0.899, 0.862]
```

`pose, fp32 topk` is the same export with float16 `aten.topk` decomposed into a float32 `topk` (values cast back to float16). The rewrite fixes your pose model on this lane, so the packed indices may well be what you see on the iPhone. I could not test that iPhone. It does not explain why the detect control is unaffected. One hint: whether the packing shows up depends on the surrounding graph. In a variant that gathers rows with the indices, appending `* 2.0` to the gathered rows made the Neural Engine indices correct.

I filed this as FB24904516.
